src/charm.py

```diff
@@ -244,37 +244,50 @@ def _command(self) -> str:
+        retention_size_setpoint = config.get("maximum_retention_size", "")
 
-        except ValueError as e:
-            logger.warning(e)
-            self.unit.status = BlockedStatus(f"Invalid retention size: {e}")
+        try:
+            if retention_size_setpoint.endswith("%"):
+                ratio = self._percent_string_to_ratio(retention_size_setpoint)
 
-        else:
-            # `storage.tsdb.retention.size` uses the legacy binary format, so "GB" and not "GiB"
-            # https://github.com/prometheus/prometheus/issues/10768
-            # For simplicity, always communicate to prometheus in GiB
-            try:
-                capacity = self._convert_k8s_capacity_to_legacy_binary_gigabytes(
+                # storage.tsdb.retention.size uses the legacy binary format, so "GB" and not "GiB"
+                # https://github.com/prometheus/prometheus/issues/10768
+                # For simplicity, always communicate to prometheus in GiB
+                capacity = self._convert_capacity_str_to_legacy_binary_gigabytes(
                     self._get_pvc_capacity(), ratio
                 )
-            except ValueError as e:
-                self.unit.status = BlockedStatus(f"Error calculating retention size: {e}")
-            except LightkubeApiError as e:
-                self.unit.status = BlockedStatus(
-                    f"Error calculating retention size "
-                    f"(try running `juju trust` on this application): {e}"
-                )
             else:
-                logger.debug("Retention size limit set to %s (%s%%)", capacity, ratio * 100)
-                args.append(f"--storage.tsdb.retention.size={capacity}")
+                # Retention size is given as an absolute value
+                ratio = 1.0
+                capacity = self._convert_capacity_str_to_legacy_binary_gigabytes(
+                    retention_size_setpoint, 1.0
+                )
+
+        except ValueError as e:
+            logger.warning(e)
+            self.unit.status = BlockedStatus(f"Error calculating retention size: {e}")
+        except LightkubeApiError as e:
+            logger.warning(e)
+            self.unit.status = BlockedStatus(
+                f"Error calculating retention size "
+                f"(try running `juju trust` on this application): {e}"
+            )
+        else:
+            logger.info(
+                "Retention size limit set to legacy binary notation '%s' "
+                "(converted from user input: '%s')",
+                capacity,
+                retention_size_setpoint,
+            )
+            args.append(f"--storage.tsdb.retention.size={capacity}")
 
         command = ["/bin/prometheus"] + args
 
         return " ".join(command)
 
-    def _convert_k8s_capacity_to_legacy_binary_gigabytes(self, capacity: str, multiplier) -> str:
+    def _convert_capacity_str_to_legacy_binary_gigabytes(
+        self, capacity: str, multiplier: float
+    ) -> str:
```

Comment:
> Shouldn't we warn the user that is trying to set the maximum_retention_size to a value that it's greater than the PVC capacity?

That's a good point. I should probably add a warning. Unless we choose to clamp the retention size to `min(pvc, user_setpoint)`, but that would be inconvenient with hostpath, because host storage can be more than the pvc.

> shouldn't these methods (_convert_capacity_str_to_legacy_binary_gigabytes, _get_pvc_capacity, _percent_string_to_ratio) be free functions? Or even be part of obsevability_libs...?

- `_get_pvc_capacity` uses `self` a bit
- the others can be static or free. Not so sure about o11y_libs.

Reply:
I'm thinking that if the user try to set  `maximum_retention_size` to a value that is greater than the PVC we should raise an error rather than a warning....

What do you think @simskij @rbarry82 ?

Reply:
I would definitely say error.

Does Prometheus "know better?" This would require testing with a CSI which can limit space, but I'd be concerned about Prometheus crashing if the size exceeds the available size. `HostPath` should be considered, basically, `--unsafe|--force`. If the claimed retention size is larger than the PVC, error.

Unless we plan to use these elsewhere, I don't know that they belong in observability libs. Honestly, I'd be happier to see `bitmath` dropped completely in favor of "hey, give sizes in GB/MB, not GiB/MiB", and those can be trivially parsed into whatever we use.

It's already there, so we have it, but it's big argument to keep this far away from observability-libs if all of a sudden you now need to `fetch-lib` _and_ get something from Pypi.

Reply:
bitmath is need for parsing the capacity strings coming from kubernetes, which can be binary/decimal/unitless.
See https://github.com/prometheus/prometheus/issues/10768.

Reply:
What I meant was "we are bringing in an outside package for something which can be done in less than 15 lines of Python".

It isn't that we don't have to convert, it's that `bitmath` itself is a single file which is <400 lines once comments/docstrings are removed, and even most of **that** is "handle this ambiguous case like whitespace between ${num} and ${unit}", which k8s will never give back to us.